[Dataset hub] more than storage: streaming, editing, connectors (#2052)
huggingface/hub-docs: docs/hub/datasets-editing.md

# Editing datasets

The [Hub](https://huggingface.co/datasets) enables collaborative curation of community and research datasets. We encourage you to explore the datasets available on the Hub and contribute to their improvement to help grow the ML community and accelerate progress for everyone. All contributions are welcome!

Start by [creating a Hugging Face Hub account](https://huggingface.co/join) if you don't have one yet.

## Edit using the Hub UI

> [!WARNING]
> This feature is only available for CSV datasets for now.

> The Hub's web interface allows users without any developer experience to edit a dataset.

Reviewer: ```suggestion
The Hub's web interface allows users without any technical expertise to edit a dataset.
```

_developer experience_ was a bit confusing to me. I think technical expertise is probably closer to what we are trying to convey

For example, [`samsum`](https://huggingface.co/datasets/knkarthick/samsum?library=datasets) shows how to do so with 🤗 Datasets below.

<div class="flex justify-center">
<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-usage.png"/>
<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-usage-dark.png"/>
</div>

<div class="flex justify-center">
<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-usage-modal.png"/>
<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/datasets-usage-modal-dark.png"/>
</div>

### Only upload the new data

Hugging Face's storage uses [Xet](https://huggingface.co/docs/hub/en/xet) which is based on deduplication, and enables in particular deduplicated uploads.
Unlike regular cloud storages, Xet doesn't require datasets to be completely reuploaded to commit changes.

> Instead, it automatically detects which parts of the dataset changed and tells the client library to only upload the parts that changed.

Reviewer 2: ```suggestion
Instead, it automatically detects which parts of the dataset have changed and instructs the client library only to upload the updated parts.
```

To do that, Xet uses a smart algorithm to find chunks of 64kB that already exist on Hugging Face.

Here is how it works with Pandas:

```python
import pandas as pd

# Load the dataset
df = pd.read_csv(f"hf://datasets/{repo_id}/data.csv")

# Edit the dataset
# df = df.apply(...)

# Commit the changes
df.to_csv(f"hf://datasets/{repo_id}/data.csv")
```

This code first loads a dataset and then edits it.
Once the edits are done, `to_csv()` materializes the file in memory, chunks it, and asks Xet which chunks are already on Hugging Face and which chunks changed, and finally only upload the new data.

### Optimized Parquet editing

Therefore the amount of data to reupload depends on the edits and the file structure.

The Parquet format is columnar and compressed at the page level (pages are around ~1MB).

> We optimized Parquet for Xet with [Parquet Content Defined Chunking](https://huggingface.co/blog/parquet-cdc), which ensures unchanged data generally result in unchanged pages.

Reviewer 3: add a visual of how those files and/or datasets are marked on the Hub 😁
